main/src/main/scala/sbt/internal/nio/CheckBuildSources.scala

```diff
@@ -8,56 +8,153 @@
+    val res = !commands.exists { c =>
+      c == LoadProject || c == RebootCommand || c == TerminateAction || c == "shutdown"
+    }
+    if (!res) {
+      previousStamps.set(getStamps(force = true))
+      needUpdate.set(false)
+    }
+    res
+  }
+  @inline private def forceCheck = fileTreeRepository.isEmpty
+  private[sbt] def needsReload(state: State, cmd: String) = {
+    (needCheck(cmd) && (forceCheck || needUpdate.compareAndSet(true, false))) && {
+      val extracted = Project.extract(state)
+      val onChanges = extracted.get(Global / onChangedBuildSource)
+      val logger = state.globalLogging.full
+      val current = getStamps(force = false)
+      val previous = previousStamps.getAndSet(current)
+      Settings.changedFiles(previous, current) match {
+        case fileChanges @ FileChanges(created, deleted, modified, _) if fileChanges.hasChanges =>
+          val rawPrefix = s"build source files have changed\n" +
+            (if (created.nonEmpty) s"new files: ${created.mkString("\n  ", "\n  ", "\n")}"
+             else "") +
+            (if (deleted.nonEmpty)
+               s"deleted files: ${deleted.mkString("\n  ", "\n  ", "\n")}"
+             else "") +
+            (if (modified.nonEmpty)
+               s"modified files: ${modified.mkString("\n  ", "\n  ", "\n")}"
+             else "")
+          val prefix = rawPrefix.linesIterator.filterNot(_.trim.isEmpty).mkString("\n")
+          if (onChanges == ReloadOnSourceChanges) {
+            logger.info(s"$prefix\nReloading sbt...")
+            true
+          } else {
+            val tail = "Apply these changes by running `reload`.\nAutomatically reload the " +
+              "build when source changes are detected by setting " +
+              "`Global / onChangedBuildSource := ReloadOnSourceChanges`.\nDisable this " +
+              "warning by setting `Global / onChangedBuildSource := IgnoreSourceChanges`."
+            logger.warn(s"$prefix\n$tail")
+            false
+          }
+        case _ => false
+      }
+    }
+  }
+  override def close(): Unit = {}
+}
 
 private[sbt] object CheckBuildSources {
+  private[sbt] val CheckBuildSourcesKey =
+    AttributeKey[CheckBuildSources]("check-build-source", "", KeyRanks.Invisible)
+  /*
+   * Reuse the same instance of CheckBuildSources across reloads but reset the state. This
+   * should allow the `set` command to work with checkBuildSources / fileInputs and
+   * checkBuildSources / pollInterval. The latter makes it possible to switch between
+   * the asynchronous and polling implementations during the same sbt session.
+   */
+  private[sbt] def init(state: State): State = state.get(CheckBuildSourcesKey) match {
+    case Some(cbs) =>
+      cbs.reset(state)
+      state
+    case _ =>
+      val cbs = new CheckBuildSources
+      cbs.reset(state)
+      state.put(CheckBuildSourcesKey, cbs)
+  }
   private[sbt] def needReloadImpl: Def.Initialize[Task[StateTransform]] = Def.task {
```

Comment:
Is this task definition still needed if you're basically doing everything with state put and get?

Reply:
Not really. I left it for bincompat.

Reply:
ok. Then could you deprecate the impl and the key plz? Otherwise, I think it's all cool.

Reply:
It actually doesn't make sense to deprecate the key because the build inputs are scoped to that key. Supposed I have some weird project layout like
```
> project
  > project-srcs
     > Build.scala
```
then I can make build monitoring work by adding
```
Global / checkBuildSources / fileInputs += baseDirectory.toGlob / "project" / "project-srcs" / "*.scala"
```
I now remember that I was thinking that if the task could be scoped, it probably should have an implementation, even if it isn't directly used by sbt. That's why I left the definition in.

Reply:
ok. That makes sense.